content/blog/2018/06/2018-06-15-remoting-over-message-bus.adoc

```diff
@@ -0,0 +1,88 @@
+
+* Kafka itself is not a queue like ActiveMQ or RabbitMQ, it is a distributed, replicated commit log. This helps to remove message delivery complexity we have in traditional queue system.
+* We need to support data streaming as a requirement, and Kafka is good at this aspect, which RabbitMQ is lack of.
+* Kafka is said to have a better scalability and good support from the development community.
+
+=== Current State
+The project is reaching the end of the first phase and here are things we have achieved so far:
+
+* Setup project as a set of Docker Compose components: Kafka cluster, Jenkins master (with plugin) and a custom agent (JAR).
+* Create a PoC with new command transport implementation to support Kafka, which involves of command invocation, RMI, classloading and data streaming.
+* Make neccessary changes in Remoting and Jenkins core to make them extensible for the use of this project.
+* Decide to use Kafka as a suitable final implementation.
+
+We planned to release an alpha version of this plugin by the end of this phase, but decided to move this release to the second phase because we need to wait for remoting and core patches to be released.
+
+=== Architecture Overview
+The project consists of multiple components:
+
+* Kafka Client Library - new command transport implementation, producer and consumer client logic.
+* Remoting Kafka Plugin - plugin implementation with KafkaGlobalConfiguration and KafkaComputerLauncher.
+* Remoting Kafka Agent - A custom JAR agent with remoting JAR packaged together with a custom Engine implementation to setup a communication channel with Kafka.
+* All the components are packaged together with Docker Compose.
+
+The below diagram is the overview of the current architecture:
+image:/images/post-images/remoting-kafka-intro/remoting-kafka-architecture.png[title="Remoting Kafka Plugin Current Architecture", role="center"]
+
+With this design, master is not communicating with agent using direct TCP communication anymore, all the communication commands are transfered with Kafka.
+
+=== Features
+
+==== 1. Kafka Global Configuration
+image:/images/post-images/remoting-kafka-intro/kafka-global-config.png[title="Remoting Kafka Global Configuration", role="center"]
+
+==== 2. Custom agent start up as a JAR
+User can start running an agent with the following command:
+image:/images/post-images/remoting-kafka-intro/start-agent.png[title="Command to start an agent", role="center"]
+
+==== 3. Launch agents with Kafka
+image:/images/post-images/remoting-kafka-intro/launch-agent-kafka.png[title="Launch agent with Kafka", role="center"]
+
+==== 4. Commands transferred between master and agent over Kafka
+image:/images/post-images/remoting-kafka-intro/kafka-commands.png[title="Master-agent communication with Kafka", role="center"]
```

Comment:
I would add some text here.
E.g. "Remoting operations are being executed over kafka. In the log you may see command execution (SlaveInstallerFactoryImpl.isWindows()), classloading (Classloader.fetch()), and log streaming (Pipe.chunk())"